### PR TITLE
[TRTLLM-12720][feat] Support nvfp4 w4a16 on sm120

### PR DESCRIPTION
**Pull request by pamelap-nvidia**
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Dev Engineer Review

The changes add SM120/121 PyTorch support for W4A16 NVFP4 inference across dense linear, LM-head, and MoE paths.

- Adds CUDA-core, Triton, Marlin, CUTLASS, and CuteDSL backend support.
- Updates Marlin build and runtime guards for SM90 and SM120.
- Adds compressed-tensors and ModelOpt configuration handling.
- Adds checkpoint mapping for packed weights and global scales.
- Adds backend-specific derived-state caching and Marlin weight transformation.
- Updates MoE backend selection and W4A16 NVFP4 scale handling.
- Restricts ReLU2 fusion to eligible static NVFP4 activation inputs.
- Clarifies NVFP4 activation-quantization checks in DeepSeek models.
- Adds SM12x support to Marlin GEMM test gating.

The reported local validation passes 74 merge-focused tests and 69 W4A16/model regression tests. Six CI runs failed or had L0 pipeline failures. Failure analysis and a new NVIDIA-triggered run are required.

## QA Engineer Review

The PR adds test functions and coverage for:

- Nemotron H MoE, MTP, and checkpoint mapping.
- CuteDSL B12x W4A16 NVFP4 backend selection and scale transformation.
- W4A16 NVFP4 linear, Marlin, Triton, LM-head, fusion, shape, and quantization behavior.
- Model configuration and compressed-tensors parsing.
- Marlin SM90/SM12x GEMM gating.
- MoE activation-quantization behavior.

No files under `tests/integration/test_lists/`, `test-db/`, `qa/`, or `waives.txt` were modified. Coverage registration for the added test functions is not shown.

Verdict: needs follow-up.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Summary                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                            
Adds end-to-end support for W4A16 NVFP4 checkpoints (packed FP4 weights + per-group FP8                                                                                                                                                                     
scales, **high-precision activations**) in the PyTorch backend, targeting Qwen3.5/3.6 and                                                                                                                                                                   
Nemotron-H NVFP4 exports on SM120/121 (RTX 5090/6000, GB10) and Hopper.                                                                                                                                                                                     
                                                                                                                                                                                                                                                            
The core problem this solves: TRT-LLM previously conflated "has NVFP4 weights" with
"quantizes activations to NVFP4". W4A16 breaks that assumption — it has FP4 weights but
consumes BF16/FP16 activations and has no calibrated `input_scale`. Every FP4 fusion site   
keyed on `has_nvfp4` would therefore feed an `Fp4QuantizedTensor` into a layer that can't
consume one, or read a `None` activation scale.
                                                                                                                              
## Changes   

**Activation-quantization capability, made explicit**                                                                         
                                                                                                                              
- New `LinearMethodBase.quantizes_nvfp4_activations` ClassVar, surfaced as          
  `Linear.has_nvfp4_activation_quantization` and `MoE.has_nvfp4_activation_quantization`.                                                                                         
                                                                                                                              
**Linear**                                         
                                                                                                                              
- `W4A16NVFP4LinearMethod`: on-the-fly Triton weight dequant. Leaves `input_scale` /
  `inv_input_scale` / `alpha` as `None` from `create_weights` onward; the parent        
  `process_weights_after_loading_*` now skip scales whose destination was never allocated.
- `MarlinNVFP4LinearMethod`: Marlin-backed W4A16, selected by default for BF16 on
  SM90/120/121. Handles its own K->64 / N->128 padding.
- Per-block scales stay in swizzled layout; the unswizzled view is a derived non-persistent
  buffer, which also makes the GMS `cache_derived_state` refresh idempotent (the previous
  in-place unswizzle would double-apply on re-materialization). 

**MoE**

- `CuteDslB12xFusedMoE` accepts `W4A16_NVFP4` and stays on the b12x path for both prefill
  and decode (`quant_mode="w4a16"`); falls back to CUTLASS for EP / attention-DP or
  non-SM120/121.
- `W4A16NVFP4CutlassFusedMoEMethod` on the CUTLASS path; shared per-layer quant-config
  lookup so mixed-precision checkpoints resolve expert configs correctly.
- Nemotron-H: promotes experts to W4A4 NVFP4 on non-CuteDSL backends, and routes BF16 MTP
  body sublayers to CUTLASS.

**C++**

- Build the dense Marlin NVFP4 GEMM/repack kernels for `120f` in addition to `90`; relax the
  runtime SM asserts to `9.x || 12.x`. No kernel logic changes. 

**No LLM API surface change** — `llm_args.py` is untouched, so no golden-manifest
regeneration.

## Behavior changes for existing users

- **Hopper NVFP4 BF16 linears now use the Marlin GEMM backend by default** (previously opt-in
  via `nvfp4_gemm_config.allowed_backends: [..., marlin]`). Attention FP4 output and the
  torch.compile inplace op are disabled on those layers as a consequence.
- MoE backend `AUTO` resolves to `CUTEDSL` for W4A16 NVFP4 on SM120/121.

## Test coverage

`test_w4a16_nvfp4_linear.py` (new) covers method selection across SM90/100/120/121 and dtype,
the Marlin-vs-Triton numerical match on GPU (including K/N padding shapes), scale-cache buffer
semantics, the mixed-precision `use_fused_gemm_allreduce` rebind, and every fusion gate that
changed. Plus new checkpoint-parsing tests (`test_model_config.py`,
`test_quant_config_utils.py`, `test_kv_cache_dtype_override.py`), Nemotron-H MoE/MTP config
tests, the weight-mapper tests, and W4A16 cases added to the b12x MoE backend suite.


## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py [hunk 1]
```diff
@@ -53,9 +53,27 @@ def _split_mamba2_mixer_in_proj(w: torch.Tensor) -> torch.Tensor:
         # input_scale) do not and are left alone.
         d_in_proj = 2 * d_inner + 2 * n_groups * d_state + nheads
 
+        def _invert_compressed_tensors_scale(value) -> torch.Tensor:
+            value = value[...] if not isinstance(value, torch.Tensor) else value
+            value = value.to(torch.float32)
+            return torch.where(value > 0, value.reciprocal(),
+                               torch.zeros_like(value)).contiguous()
+
+        def _canonicalize_quant_weight(key: str, value):
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: please  add a comment about why and which quantization config will use the `if condition`.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py [hunk 2]
```diff
@@ -162,7 +164,26 @@ def forward(
...
+        model_config: ModelConfig[PretrainedConfig],
+        layer_idx: int) -> QuantConfig | None:
+    # Per-expert mixed precision config is more specific than the global config.
+    layer_quant_config = _get_layer_quant_config(model_config, layer_idx,
+                                                 "mixer.experts")
+    if layer_quant_config is not None:
+        return layer_quant_config
+
+    quant_config = model_config.quant_config
+    if (quant_config is not None and quant_config.quant_algo
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_nemotron_h.py]: it is quite confusing to me.

what if we pass MARLIN moe backend on Hopper, would it fallback to NVFP4 will cause issue?

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py (hunk 1 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: I removed this part as it's not needed.

File: tensorrt_llm/_torch/models/modeling_nemotron_h.py (hunk 2 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/models/modeling_nemotron_h.py]: Good point. I removed this fallback.

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def _split_mamba2_mixer_in_proj(w: torch.Tensor) -> torch.Tensor:
         new_weights = {}
         for name, _ in weights.items():
             key = name
+            value = weights[name]
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: depends on your choice, but seems this file changes are not necessary.

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py [hunk 4]
```diff
@@ -27,6 +27,24 @@
...
+def _get_layer_quant_config(
+        model_config: ModelConfig,
+        layer_idx: Optional[int],
+        expert_module_suffix: str = "mlp.experts") -> Optional[QuantConfig]:
+    if layer_idx is None or model_config.quant_config_dict is None:
+        return None
+
+    moe_module_name = f"model.layers.{layer_idx}.{expert_module_suffix}"
+    for name, quant_config in model_config.quant_config_dict.items():
+        if _is_same_or_child_module_path(name, moe_module_name):
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: would `model.layers.5*` cause bug?

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py [hunk 5]
```diff
@@ -78,17 +96,28 @@ def get_moe_cls(
...
-            if quant_config.quant_mode.has_nvfp4():
+            has_nvfp4 = (quant_config.quant_mode.has_nvfp4()
+                         and not has_w4a16_nvfp4)
+            if has_nvfp4 or has_w4a16_nvfp4:
                 from tensorrt_llm._utils import get_sm_version
                 sm_version = get_sm_version()
                 if sm_version in CuteDslB12xFusedMoE._SUPPORTED_SM_VERSIONS:
+                    mapping = model_config.mapping
+                    if mapping.moe_ep_size > 1 or mapping.dp_size > 1:
+                        logger.info(
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: logger.warning might be more proper.

File: tensorrt_llm/_torch/model_config.py [hunk 6]
```diff
@@ -351,6 +351,18 @@ def resolve_moe_backend(moe_backend: str,
             sm_version = get_sm_version()
             if 100 <= sm_version < 120:
                 return "TRTLLM"
+            return "CUTLASS"
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/model_config.py]: curious why adding this fallback? seems not related with your code changes

File: tests/unittest/_torch/models/checkpoints/hf/test_nemotron_h_weight_mapper.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (Wanli-Jiang) [inline tests/unittest/_torch/models/checkpoints/hf/test_nemotron_h_weight_mapper.py]: seems this test is not needed actually.

**Comment by juney-nvidia** [comment]
@rosong11  Song can you help review this PR on behalf from trt-llm-moe-devs and trt-llm-kernels-devs modules?
@yuxianq  Yuxian can you help coordinate the review from attention module perspective? 
@QiJune  Jun pls help coordinate the review from runtime and baseline perspective. 
@VALLIS-NERIA  Xiwen pls help review from KV Cache Manager perspective.

File: tensorrt_llm/_torch/model_config.py [hunk 8]
```diff
@@ -351,6 +351,18 @@ def resolve_moe_backend(moe_backend: str,
...
+        is_w4a16_nvfp4 = (quant_config is not None and quant_config.quant_algo
+                          in (QuantAlgo.W4A16_NVFP4, "W4A16_NVFP4"))
+        if is_w4a16_nvfp4:
+            sm_version = get_sm_version()
+            # CuteDslB12xFusedMoE on SM120/121, MarlinFusedMoE on Hopper. Any
+            # other SM falls through to CUTLASS, which dequantizes on the fly.
+            if sm_version in (120, 121):
+                return "CUTEDSL"
+            if 90 <= sm_version < 100:
+                return "MARLIN"
```
Comment (farazkh80) [inline tensorrt_llm/_torch/model_config.py:364]: AFAIU, auto resolves w4a16 to MARLIN on Hopper, but `MarlinFusedMoE.can_implement` rejects it (never called on this path) and `_get_quant_method` returns the plain-NVFP4 method which reads input_scale/alpha a W4A16 checkpoint doesn't have. Would this be an issue?

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py [hunk 9]
```diff
@@ -1,3 +1,18 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (farazkh80) [inline tests/unittest/_torch/thop/parallel/test_fp4_linear.py:1]: should we add these to l0 for CI?

**Comment by farazkh80** [comment]
Hey @pamelap-nvidia, the PR description says:

>   Checkpoint parsing
> - quant_config_utils.py: compressed-tensors configs with input_activations: null + format: nvfp4-pack-quantized + 4-bit float weights at group_size=16 now resolve to QuantAlgo.W4A16_NVFP4 instead of raising.
>  - model_config.py / llm_utils.py: when hf_quant_config.json declares NVFP4 but the inline config.json compressed-tensors block says weight-only, the inline activation semantics win.

but I couldn't find these in the diff (`quant_config_utils.py` / `llm_utils.py` aren't among the changed files) — was this split into a separate PR?

**Comment by pamelap-nvidia** [comment]
> Hey @pamelap-nvidia, the PR description says:
> 
> > Checkpoint parsing
> > 
> > * quant_config_utils.py: compressed-tensors configs with input_activations: null + format: nvfp4-pack-quantized + 4-bit float weights at group_size=16 now resolve to QuantAlgo.W4A16_NVFP4 instead of raising.
> > * model_config.py / llm_utils.py: when hf_quant_config.json declares NVFP4 but the inline config.json compressed-tensors block says weight-only, the inline activation semantics win.
> 
> but I couldn't find these in the diff (`quant_config_utils.py` / `llm_utils.py` aren't among the changed files) — was this split into a separate PR?

I removed the relevant changes as they're no longer needed.

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py (hunk 9 above)
Comment (pamelap-nvidia) [reply tests/unittest/_torch/thop/parallel/test_fp4_linear.py:1]: Added to DGX Spark test. Hopper test already includes the parent folder.

File: tensorrt_llm/_torch/model_config.py (hunk 8 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/model_config.py:364]: Updated can_implement.

File: tests/unittest/_torch/models/checkpoints/hf/test_nemotron_h_weight_mapper.py (hunk 7 above)
Comment (pamelap-nvidia) [reply tests/unittest/_torch/models/checkpoints/hf/test_nemotron_h_weight_mapper.py]: Makes sense. Removed.

File: tensorrt_llm/_torch/model_config.py (hunk 6 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/model_config.py]: Makes sense. Removed.

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py (hunk 5 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: Done.

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py (hunk 3 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: Makes sense. Removed.

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py (hunk 4 above)
Comment (pamelap-nvidia) [reply tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: I reverted it to the original format. From manual runs it's not a concern yet.

**Comment by schetlur-nv** [comment]
/bot skip --comment "Failure analysis:  the failing SBSA multi-GPU Slurm job completed successfully with exit code 0:0; the pipeline failed in CI Slurm log-tracking/teardown after tail -f lost its transport endpoint and kill $tailPid errored. https://pbss.s8k.io/v1/AUTH_svc_tensorrt/sw-tensorrt-ci-analysis/LLM/main/L0_MergeRequest_PR/52563/failure_analysis.html "
